applications/tari_indexer/src/event_scanner.rs

```diff
@@ -0,0 +1,561 @@
+            })
+            .await?;
+
+        match PayloadResultStatus::try_from(response.status) {
+            Ok(PayloadResultStatus::Finalized) => {
+                let proto_decision = tari_dan_p2p::proto::consensus::Decision::try_from(response.final_decision)?;
+                let final_decision = proto_decision.try_into()?;
+                if let Decision::Commit = final_decision {
+                    Ok(Some(response.execution_result)
+                        .filter(|r| !r.is_empty())
+                        .map(|r| decode(&r))
+                        .transpose()?)
+                } else {
+                    Ok(None)
+                }
+            },
+            _ => Ok(None),
+        }
+    }
+
+    fn extract_events_from_transaction_result(&self, result: ExecuteResult) -> Vec<EventData> {
+        if let TransactionResult::Accept(substate_diff) = result.finalize.result {
+            let substates: HashMap<SubstateId, Substate> = substate_diff.into_up_iter().collect();
+
+            let events = result
+                .finalize
+                .events
+                .into_iter()
+                .map(|event| {
+                    let substate = if let Some(substate_id) = event.substate_id() {
+                        substates.get(&substate_id).cloned()
+                    } else {
+                        None
+                    };
+
+                    EventData { event, substate }
+                })
+                .collect();
+
+            events
+        } else {
+            vec![]
+        }
+    }
+
+    fn extract_transaction_ids_from_blocks(&self, blocks: Vec<Block>) -> HashSet<TransactionId> {
+        let mut transaction_ids = HashSet::new();
+
+        for block in blocks {
+            for command in block.commands() {
+                match command {
+                    Command::Accept(t) | Command::LocalOnly(t) => {
+                        transaction_ids.insert(*t.id());
+                    },
+                    _ => {
+                        // we are only interested in events from confirmed transactions
+                    },
+                }
+            }
+        }
+
+        transaction_ids
+    }
+
+    fn build_genesis_block_id(&self) -> BlockId {
+        let start_block = Block::zero_block(self.network);
+        *start_block.id()
+    }
+
+    #[allow(unused_assignments)]
+    async fn get_new_blocks_from_committee(
+        &self,
+        shard: Shard,
+        committee: &mut Committee<PeerAddress>,
+        epoch: Epoch,
+    ) -> Result<Vec<Block>, anyhow::Error> {
+        // We start scanning from the last scanned block for this commitee
+        let start_block_id = {
+            let mut tx = self.substate_store.create_read_tx()?;
+            tx.get_last_scanned_block_id(epoch, shard)?
+        };
+        let start_block_id = start_block_id.unwrap_or(self.build_genesis_block_id());
+
+        committee.members.shuffle(&mut OsRng);
```

Comment:
`Committee` has a shuffle method. In future, we'll want to remove the `pub` from the `members` field
```suggestion
        committee.shuffle();
```